File: optik-platform/backend/utils/mongo_init.py
"""
MongoDB Initialization and Startup/Shutdown Hooks
Integrates MongoDB with FastAPI application lifecycle
"""

import logging
from fastapi import FastAPI
from utils.mongo_db import mongodb, JobRepository, ProductRepository
from utils.env import allow_demo_data
from datetime import datetime
import uuid

logger = logging.getLogger(__name__)


async def init_mongodb(app: FastAPI):
    """
    Initialize MongoDB connection on app startup.

    Usage in main.py:
        @app.on_event("startup")
        async def startup():
            await init_mongodb(app)
    """
    logger.info("🔄 Initializing MongoDB...")

    # Connect to MongoDB
    connected = await mongodb.connect()

    if not connected:
        logger.error("❌ Failed to connect to MongoDB on startup")
        raise RuntimeError("MongoDB connection failed")

    logger.info("✅ MongoDB initialized successfully")

    # Seed initial data if collections are empty
    await seed_initial_data()

    # Store mongodb in app state for access
    app.state.mongodb = mongodb


async def close_mongodb(app: FastAPI):
    """
    Close MongoDB connection on app shutdown.

    Usage in main.py:
        @app.on_event("shutdown")
        async def shutdown():
            await close_mongodb(app)
    """
    logger.info("🔄 Closing MongoDB connection...")
    await mongodb.disconnect()
    logger.info("✅ MongoDB closed")


async def seed_initial_data():
    """
    Seed initial data if collections are empty.
    Useful for development and testing.
    """
    if not allow_demo_data():
        logger.info("Demo data seeding disabled outside non-production environments.")
        return
    try:
        # Check if collections are empty
        job_repo = JobRepository()
        product_repo = ProductRepository()

        job_count = await job_repo.count()
        product_count = await product_repo.count()

        if job_count == 0:
            logger.info("📊 Seeding initial jobs data...")
            await seed_jobs()

        if product_count == 0:
            logger.info("📊 Seeding initial products data...")
            await seed_products()

        logger.info("✅ Initial data seeding complete")

    except Exception as e:
        logger.warning(f"⚠️ Could not seed initial data: {str(e)}")


async def seed_jobs():
    """Seed initial jobs for testing."""
    job_repo = JobRepository()

    initial_jobs = [
        {
            "id": str(uuid.uuid4()),
            "user_id": "user_demo_001",
            "store_url": "https://demo-store.shopify.com",
            "platform": "shopify",
            "tier": "growth",
            "status": "completed",
            "message": "Store conversion completed successfully",
            "deployment_config": {
                "network": "devnet",
                "treasury": "7dzzihnceMRrhDvDFVH5E4pVKhFeEgLTTfxWczvbHrPa"
            },
            "created_at": datetime.utcnow(),
            "updated_at": datetime.utcnow()
        },
        {
            "id": str(uuid.uuid4()),
            "user_id": "user_demo_002",
            "store_url": "https://example-woo.com",
            "platform": "woocommerce",
            "tier": "scale",
            "status": "processing",
            "message": "Analyzing store products...",
            "deployment_config": None,
            "created_at": datetime.utcnow(),
            "updated_at": datetime.utcnow()
        }
    ]

    for job in initial_jobs:
        await job_repo.create(job)

    logger.info(f"✅ Seeded {len(initial_jobs)} jobs")


async def seed_products():
    """Seed initial products for testing."""
    product_repo = ProductRepository()

    initial_products = [
        {
            "id": str(uuid.uuid4()),
            "user_id": "user_demo_001",
            "name": "Genesis Hoodie",
            "description": "Limited edition founding member hoodie",
            "supply": "100",
            "sold": 45,
            "price": "0.5 SOL",
            "status": "Live",
            "created_at": datetime.utcnow(),
            "updated_at": datetime.utcnow()
        },
        {
            "id": str(uuid.uuid4()),
            "user_id": "user_demo_001",
            "name": "Founder Cap",
            "description": "Exclusive founder cap with blockchain badge",
            "supply": "50",
            "sold": 50,
            "price": "1.2 SOL",
            "status": "Sold Out",
            "created_at": datetime.utcnow(),
            "updated_at": datetime.utcnow()
        },
        {
            "id": str(uuid.uuid4()),
            "user_id": "user_demo_001",
            "name": "Digital Badge",
            "description": "Unlimited digital membership badge",
            "supply": "Unlimited",
            "sold": 1240,
            "price": "0.1 SOL",
            "status": "Live",
            "created_at": datetime.utcnow(),
            "updated_at": datetime.utcnow()
        }
    ]

    for product in initial_products:
        await product_repo.create(product)

    logger.info(f"✅ Seeded {len(initial_products)} products")


async def health_check():
    """Get health status of MongoDB."""
    return await mongodb.health_check()

File: optik-platform/backend/scripts/mongo-init.js
/**
 * MongoDB Initialization Script
 * Runs on container startup to set up users and collections
 */

const env = (typeof process !== "undefined" && process.env) ? process.env : {};
const environment = (env.ENVIRONMENT || "production").toLowerCase();
const allowDemo = env.OPTIK_ALLOW_DEMO_DATA === "true" && environment !== "production";

const appUser = env.OPTIK_MONGO_APP_USER || (allowDemo ? "optik_user" : null);
const appPassword = env.OPTIK_MONGO_APP_PASSWORD || (allowDemo ? "optik_password_change_me" : null);

if (!appUser || !appPassword) {
  throw new Error("OPTIK_MONGO_APP_USER and OPTIK_MONGO_APP_PASSWORD are required to initialize MongoDB.");
}

// Switch to admin database
db = db.getSiblingDB('admin');

// Create Optik application user (if not already exists)
try {
  db.createUser({
    user: appUser,
    pwd: appPassword,
    roles: [
      { role: "readWrite", db: "optik" },
      { role: "dbOwner", db: "optik" }
    ]
  });
  print("✅ Created optik_user");
} catch (e) {
  print("⚠️ User might already exist: " + e);
}

// Switch to optik database
db = db.getSiblingDB('optik');

// Create collections with schema validation
print("📚 Creating collections...");

// Jobs Collection
db.createCollection("jobs", {
  validator: {
    $jsonSchema: {
      bsonType: "object",
      required: ["id", "user_id", "store_url", "platform", "status"],
      properties: {
        id: { bsonType: "string", description: "Unique job identifier" },
        user_id: { bsonType: "string", description: "User ID" },
        store_url: { bsonType: "string", description: "Store URL" },
        platform: {
          enum: ["shopify", "woocommerce", "custom"],
          description: "E-commerce platform"
        },
        tier: {
          enum: ["basic", "growth", "global", "scale", "elite"],
          description: "Service tier"
        },
        status: {
          enum: ["pending", "processing", "completed", "failed"],
          description: "Job status"
        },
        message: { bsonType: "string" },
        error: { bsonType: ["string", "null"] },
        deployment_config: { bsonType: ["object", "null"] },
        created_at: { bsonType: "date" },
        updated_at: { bsonType: "date" }
      }
    }
  }
});
print("✅ Created 'jobs' collection");

// Products Collection
db.createCollection("products", {
  validator: {
    $jsonSchema: {
      bsonType: "object",
      required: ["id", "user_id", "name", "price"],
      properties: {
        id: { bsonType: "string" },
        user_id: { bsonType: "string" },
        name: { bsonType: "string" },
        description: { bsonType: ["string", "null"] },
        supply: { bsonType: "string" },
        sold: { bsonType: "int" },
        price: { bsonType: "string" },
        status: {
          enum: ["Live", "Draft", "Sold Out", "Archived"],
          description: "Product status"
        },
        created_at: { bsonType: "date" },
        updated_at: { bsonType: "date" }
      }
    }
  }
});
print("✅ Created 'products' collection");

// Conversions Collection
db.createCollection("conversions");
print("✅ Created 'conversions' collection");

// Deployments Collection
db.createCollection("deployments", {
  validator: {
    $jsonSchema: {
      bsonType: "object",
      required: ["id", "job_id", "tx_hash", "network"],
      properties: {
        id: { bsonType: "string" },
        job_id: { bsonType: "string" },
        tx_hash: { bsonType: "string" },
        merchant_pda: { bsonType: "string" },
        network: {
          enum: ["devnet", "testnet", "mainnet"],
          description: "Blockchain network"
        },
        dapp_url: { bsonType: "string" },
        created_at: { bsonType: "date" }
      }
    }
  }
});
print("✅ Created 'deployments' collection");

// Integrations Collection
db.createCollection("integrations");
print("✅ Created 'integrations' collection");

// Create Indexes for Performance
print("📊 Creating indexes...");

// Jobs indexes
db.jobs.createIndex({ user_id: 1 });
db.jobs.createIndex({ status: 1 });
db.jobs.createIndex({ created_at: -1 });
db.jobs.createIndex({ user_id: 1, status: 1 });
print("✅ Created jobs indexes");

// Products indexes
db.products.createIndex({ user_id: 1 });
db.products.createIndex({ name: "text" });
db.products.createIndex({ status: 1 });
db.products.createIndex({ created_at: -1 });
print("✅ Created products indexes");

// Conversions indexes
db.conversions.createIndex({ job_id: 1 });
db.conversions.createIndex({ created_at: -1 });
print("✅ Created conversions indexes");

// Deployments indexes
db.deployments.createIndex({ job_id: 1 });
db.deployments.createIndex({ tx_hash: 1 });
db.deployments.createIndex({ merchant_pda: 1 });
db.deployments.createIndex({ network: 1 });
print("✅ Created deployments indexes");

// Integrations indexes
db.integrations.createIndex({ user_id: 1 });
db.integrations.createIndex({ name: 1 });
db.integrations.createIndex({ status: 1 });
print("✅ Created integrations indexes");

// Seed initial data (optional)
if (allowDemo) {
  print("📌 Seeding initial data (demo only)...");

  // Sample jobs
  db.jobs.insertMany([
    {
      id: "job_001",
      user_id: "user_demo",
      store_url: "https://demo.shopify.com",
      platform: "shopify",
      tier: "growth",
      status: "completed",
      message: "Store conversion completed",
      deployment_config: {
        network: "devnet",
        treasury: "7dzzihnceMRrhDvDFVH5E4pVKhFeEgLTTfxWczvbHrPa"
      },
      created_at: new Date(),
      updated_at: new Date()
    }
  ]);
  print("✅ Seeded initial data");
} else {
  print("ℹ️ Demo seed data disabled for this environment.");
}

// Setup backup index
db.adminCommand({
  createBackup: 1,
  backupName: "initial_backup"
});

print("\n" + "=".repeat(50));
print("✅ MongoDB initialization complete!");
print("=".repeat(50));
print("\nDatabase: optik");
print("Collections: jobs, products, conversions, deployments, integrations");
print("Indexes: Created for performance optimization");
print("User: optik_user created");
print("\nMongo Express: http://localhost:8081");
print("Admin: admin / admin");
